File: 700. Search in a Binary Search Tree/Solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def searchBST2(self, root, val):
        """
        :type root: TreeNode
        :type val: int
        :rtype: TreeNode
        """
        if root is None: return None
        elif root.val == val: return root
        elif root.val > val: return self.searchBST(root.left, val);
        else: return self.searchBST(root.right, val)

    def searchBST(self, root, val):
        stack = [root]
        while len(stack) != 0 and root is not None:
            node = stack.pop()
            if node is None:
                return None
            elif node.val == val: 
                return node
            elif node.val > val: 
                stack.append(node.left)
            else:
                stack.append(node.right)
        return None

File: 700. Search in a Binary Search Tree/Solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* searchBST2(TreeNode* root, int val) {
        if (!root) return NULL;
        else if (root->val == val) return root;
        else if (root->val > val) return searchBST(root->left, val);
        else return searchBST(root->right, val);
    }

    TreeNode* searchBST3(TreeNode* root, int val) {
        stack<TreeNode*> s; s.push(root);
        while (!s.empty() && root) {
            TreeNode* n = s.top(); s.pop();
            if (n == NULL) return NULL;
            else if (n->val == val) return n;
            else if (n->val > val) s.push(n->left);
            else s.push(n->right);
        }
        return NULL;
    }
    
    TreeNode* searchBST(TreeNode* root, int val) {
        TreeNode* node = root;
        while (node) {
            if (node->val == val) return node;
            else if (node->val > val) node = node->left;
            else node = node->right;
        }
        return NULL;
    }
};

